packages/opentelemetry-resources/src/platform/node/detectors/AwsBeanstalkDetector.ts

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Resource } from '../../../Resource';
+import { SERVICE_RESOURCE } from '../../../constants';
+import { Detector } from '../../../types';
+import { ResourceDetectionConfigWithLogger } from '../../../config';
+import * as fs from 'fs';
+import * as util from 'util';
+
+/**
+ * The AwsBeanstalkDetector can be used to detect if a process is running in AWS Elastic
+ * Beanstalk and return a {@link Resource} populated with data about the beanstalk
+ * plugins of AWS X-Ray. Returns an empty Resource if detection fails.
+ *
+ * See https://docs.amazonaws.cn/en_us/xray/latest/devguide/xray-guide.pdf
+ * for more details about detecting information of Elastic Beanstalk plugins
+ */
+
+// for testing purpose, build an object here
+export const cache = {
```

Comment:
Why is this exported? If you need to use these methods in the tests, you should just promisify them there. If you need to replace them for tests, they should be private members of the class.

Reply:
Thank you for pointing out! Will also modify ECS PR

Reply:
One thing here is that if I promisify method in test file, I did not find any plausible way to take use of ```sinon``` to stub only single method without wrapped with an Object.
If I make these methods private, the ```sinon.sandbox.stub``` cannot access to these methods to mock functionality.
So I made a new commit like below

Reply:
You can stub `fs.readFile` which will be called by the promisified version.

Reply:
I tried to write
```const readFileAsync = util.promisify(fs.readFile);```
outside of class, and then write
```
    readStub = sandbox
      .stub(fs, 'readFile')
      .yields(null, correctCgroupData);
```
In the test part, it did not work, wield. That is why I initially try to use a Object to wrap these functions

Reply:
@dyladan Not too sure but it seems like an ordering issue - if the methods are promisified as constants in the logic file, they'll get set, using the nodejs defaults, before the test gets a chance to stub. Is it kosher to add logic between import statements, to allow stugging between these two lines?

```
import * as sinon from 'sinon';

sinon.stub(fs.readFile)

import { awsBeanstalkDetector } from '../../src/platform/node/detectors/AwsBeanstalkDetector';
```

Otherwise, I'd generally use a comment like `// Visible for testing` when exporting something only for use in a test

Reply:
Hmm I would prefer not to rely on logic between imports. Exporting for test visibility is fine, but I want to make sure it is not exported from the module so it is not visible to end customers. I would suggest you make these methods private static members of the class, then those static members can be stubbed with sinon.

Reply:
In this case,  I tried
1. export ```AwsBeanstalkDetector``` class
2. then do ```sandbox.stub(AwsBeanstalkDetector, 'fileAccessAsync')```
It still failed in test part because ```sinon``` cannot access private member

Reply:
Yes it can, you just have to cast the type. It is a type failure, but the stub should still work.

Reply:
A dumb question, what do you mean by "cast the type"?
I did:
1. make promisify method "private static"
2.export the class ```AwsBeanstalkDetector```
3. do ```sandbox.stub(AwsBeanstalkDetector, 'fileAccessAsync')``` in test part

And this way cannot pass ```npm run compile```

Reply:
`sandbox.stub(AwsBeanstalkDetector, 'fileAccessAsync' as any)` will compile

Reply:
Amazing! It works, will also modify for ECS PR. Thank you so much for helping me out!